persistence-api/src/main/java/io/stargate/db/Persistence.java

```diff
@@ -15,96 +15,162 @@
+     * @throws AuthenticationException if the user is not authorized to login.
+     */
+    void login(AuthenticatedUser user) throws AuthenticationException;
+
+    /** The user logged in, if any. */
+    Optional<AuthenticatedUser> loggedUser();
+
+    /** Information on the client for which the connection was created, if any. */
+    Optional<ClientInfo> clientInfo();
+
+    /**
+     * The default keyspace that is in use on this connection, if any (that is, if a {@code USE}
+     * query has been performed on this connection).
+     */
+    Optional<String> usedKeyspace();
+
+    /**
+     * Prepare a query on this connection.
+     *
+     * @param query the query to prepare.
+     * @param parameters the parameters for the preparation (note that preparation only use a subset
+     *     of parameters).
+     * @return a future that, on completion of the preparation, provides its result. Please see the
+     *     {@link #execute} method for warnings regarding the completion of that future. The same
+     *     warnings apply for this method as well.
+     */
+    CompletableFuture<Result.Prepared> prepare(String query, Parameters parameters);
+
+    /**
+     * Executes a statement on this connection.
+     *
+     * @param statement the statement to execute.
+     * @param parameters the parameters for the execution.
+     * @param queryStartNanoTime the start time of the query, as a nano time returned by {@link
+     *     System#nanoTime()}.
+     * @return a future that, on completion of the execution, provides its result. Please note that
+     *     persistence implementations are allowed to complete this future on an internal, and
+     *     potentially performance sensitive, thread. As such, consumers should <b>not</b> "chain"
+     *     costly/blocking operations on this future without passing their own executor. In other
+     *     words, do not do something like: <code>
+     *       execute(...).thenRun(() -> { someBlockingOperation(); })
+     *     </code> Instead, use one of the "async" variants of {@link CompletableFuture}, namely:
+     *     <code>
+     *       // you can also provide your own executor below obviously
+     *       execute(...).thenRunAsync(() -> { someBlockingOperation(); })
+     *     </code>
+     */
+    CompletableFuture<Result> execute(
+        Statement statement, Parameters parameters, long queryStartNanoTime);
+
+    /**
+     * Executes a batch on this connection.
+     *
+     * @param batch the batch to execute.
+     * @param parameters the parameters for the batch execution.
+     * @return a future that, on completion of the execution, provides its result. Please see the
+     *     {@link #execute} method for warnings regarding the completion of that future. The same
+     *     warnings apply for this method as well.
+     */
+    CompletableFuture<Result> batch(Batch batch, Parameters parameters, long queryStartNanoTime);
+
+    void captureClientWarnings();
```

Comment:
Removing the inline request execution code is a good idea. That approach may give us another opportunity. We can probably remove these warning methods from `Persistence`. 

How about pulling the warnings from `ResultMessage` using `ResultMessage.getWarnings()`  (it's not private, so no reflection needed :partying_face:)? This might work as long as there's no warnings on `ErrorMessage`, otherwise we'd need to handle that path too.

```diff
diff --git a/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Message.java b/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Message.java
index 579e10f..5346498 100644
--- a/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Message.java
+++ b/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Message.java
@@ -658,9 +658,6 @@ public abstract class Message {
 
         final Persistence.Connection persistenceConnection = connection.persistenceConnection();
 
-        if (connection.getVersion().isGreaterOrEqualTo(ProtocolVersion.V4))
-          persistenceConnection.captureClientWarnings();
-
         connection.validateNewMessage(request.type, connection.getVersion());
 
         logger.trace("Received: {}, v={}", request, connection.getVersion());
@@ -675,7 +672,6 @@ public abstract class Message {
               } else {
                 try {
                   response.setStreamId(request.getStreamId());
-                  response.setWarnings(persistenceConnection.getClientWarnings());
                   response.attach(connection);
                   connection.applyStateTransition(request.type, response.type);
 
@@ -691,8 +687,6 @@ public abstract class Message {
                       "Failed to reply, got another error whilst writing reply: {}",
                       t.getMessage(),
                       t);
-                } finally {
-                  persistenceConnection.resetClientWarnings();
                 }
               }
             });
diff --git a/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/messages/ResultMessage.java b/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/messages/ResultMessage.java
index 29a7982..96bf109 100644
--- a/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/messages/ResultMessage.java
+++ b/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/messages/ResultMessage.java
@@ -67,6 +67,7 @@ public class ResultMessage extends Message.Response {
     super(Message.Type.RESULT);
     this.result = result;
     this.tracingId = result.getTracingId();
+    this.warnings = result.getWarnings();
   }
 
   public static class VoidSubCodec implements CBCodec<Result> {
diff --git a/persistence-api/src/main/java/io/stargate/db/Persistence.java b/persistence-api/src/main/java/io/stargate/db/Persistence.java
index 963fb22..24202da 100644
--- a/persistence-api/src/main/java/io/stargate/db/Persistence.java
+++ b/persistence-api/src/main/java/io/stargate/db/Persistence.java
@@ -166,11 +166,5 @@ public interface Persistence {
      *     warnings apply for this method as well.
      */
     CompletableFuture<Result> batch(Batch batch, Parameters parameters, long queryStartNanoTime);
-
-    void captureClientWarnings();
-
-    List<String> getClientWarnings();
-
-    void resetClientWarnings();
   }
 }
diff --git a/persistence-api/src/main/java/io/stargate/db/Result.java b/persistence-api/src/main/java/io/stargate/db/Result.java
index ca5b8e6..3ae091e 100644
--- a/persistence-api/src/main/java/io/stargate/db/Result.java
+++ b/persistence-api/src/main/java/io/stargate/db/Result.java
@@ -50,6 +50,7 @@ public abstract class Result {
 
   public final Kind kind;
   private UUID tracingId;
+  private List<String> warnings;
 
   private Result(Kind kind) {
     this.kind = kind;
@@ -64,6 +65,15 @@ public abstract class Result {
     return tracingId;
   }
 
+  public Result setWarnings(List<String> warnings) {
+    this.warnings = warnings;
+    return this;
+  }
+
+  public List<String> getWarnings() {
+    return warnings;
+  }
+
   public static class Rows extends Result {
     public final List<List<ByteBuffer>> rows;
     public final ResultMetadata resultMetadata;
diff --git a/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java b/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java
index 6ed6a08..214d658 100644
--- a/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java
+++ b/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java
@@ -433,20 +433,5 @@ public class CassandraPersistence
         return Conversion.toInternal(((BoundStatement) statement).preparedId());
       }
     }
-
-    @Override
-    public void captureClientWarnings() {
-      ClientWarn.instance.captureWarnings();
-    }
-
-    @Override
-    public List<String> getClientWarnings() {
-      return ClientWarn.instance.getWarnings();
-    }
-
-    @Override
-    public void resetClientWarnings() {
-      ClientWarn.instance.resetWarnings();
-    }
   }
 }
diff --git a/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java b/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java
index ec60cf8..1ff87b4 100644
--- a/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java
+++ b/persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java
@@ -437,7 +437,9 @@ public class Conversion {
 
   public static Result toResult(
       ResultMessage resultMessage, org.apache.cassandra.transport.ProtocolVersion version) {
-    return toResultInternal(resultMessage, version).setTracingId(resultMessage.getTracingId());
+    return toResultInternal(resultMessage, version)
+        .setTracingId(resultMessage.getTracingId())
+        .setWarnings(resultMessage.getWarnings());
   }
 
   public static Result toResultInternal(
diff --git a/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java b/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java
index a092ebd..8875647 100644
--- a/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java
+++ b/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/CassandraPersistence.java
@@ -424,20 +424,5 @@ public class CassandraPersistence
         return Conversion.toInternal(((BoundStatement) statement).preparedId());
       }
     }
-
-    @Override
-    public void captureClientWarnings() {
-      ClientWarn.instance.captureWarnings();
-    }
-
-    @Override
-    public List<String> getClientWarnings() {
-      return ClientWarn.instance.getWarnings();
-    }
-
-    @Override
-    public void resetClientWarnings() {
-      ClientWarn.instance.resetWarnings();
-    }
   }
 }
diff --git a/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java b/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java
index 919a6bf..8b4eb0b 100644
--- a/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java
+++ b/persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java
@@ -343,7 +343,8 @@ public class Conversion {
   public static Result toResult(
       ResultMessage resultMessage, org.apache.cassandra.transport.ProtocolVersion version) {
     return toResultInternal(resultMessage, version)
-        .setTracingId(ReflectionUtils.getTracingId(resultMessage));
+        .setTracingId(ReflectionUtils.getTracingId(resultMessage))
+        .setWarnings(resultMessage.getWarnings());
   }
 
   private static Result toResultInternal(
diff --git a/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java b/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java
index f2807ed..e1238b6 100644
--- a/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java
+++ b/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java
@@ -445,7 +445,9 @@ public class Conversion {
 
   public static Result toResult(
       ResultMessage resultMessage, org.apache.cassandra.transport.ProtocolVersion version) {
-    return toResultInternal(resultMessage, version).setTracingId(resultMessage.getTracingId());
+    return toResultInternal(resultMessage, version)
+        .setTracingId(resultMessage.getTracingId())
+        .setWarnings(resultMessage.getWarnings());
   }
 
   private static Result toResultInternal(
diff --git a/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java b/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java
index d26c583..e08a664 100644
--- a/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java
+++ b/persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DsePersistence.java
@@ -448,20 +448,5 @@ public class DsePersistence
         return Conversion.toInternal(((BoundStatement) statement).preparedId());
       }
     }
-
-    @Override
-    public void captureClientWarnings() {
-      ClientWarn.instance.captureWarnings();
-    }
-
-    @Override
-    public List<String> getClientWarnings() {
-      return ClientWarn.instance.getWarnings();
-    }
-
-    @Override
-    public void resetClientWarnings() {
-      ClientWarn.instance.resetWarnings();
-    }
   }
 }
```

Reply:
We could do this in a follow up ticket. It's probably more important to get this merged.

Reply:
Good idea, that's cleaner. I'll do that (but I'll deal with `ErrorMessage`; I don't know of the top of my mind if it's possible today, but in theory there is no reason a query that errors out "late" can't also have warnings so better be on the safe side).